### PR TITLE
link lost tracks to taus

### PR DESCRIPTION
**Pull request by mbluj**
#### PR description:

This PR adds links between taus and tracks used to build their charged hadron candidates, but not contained by PF-particles. The links are added at both RECO/AOD and MiniAOD levels.

At RECO/AOD level the link is created by a dedicated "modifier" plugin which adds to a tau object edm::Ptrs to members of (general)track collection, while at miniAOD level it is done via modified PATTauSlimmer module which stores in tau object edm::Ptrs to members of lostTracks collection. In both cases already existing data members of respectively reco::BaseTau and  pat::Tau objects are reused. Linking at RECO/AOD is needed to produce link at miniAOD.

Expected changes: small number of taus (<1%) will contain up to 2 additional edm::Ptrs. Other tau features will stay unchanged.    

#### PR validation:

Standard tests, i.e. runTheMatrix.py -l limited -i all --ibeos, passed successfully.

Backward-compatibility checks: updated CMSSW code reads successfully old data thanks to unchanged structure of data members of tau classes (of course vectors of Ptrs to lost tracks are always empty).

### CODE REVIEW

**Comment by jpata** [comment]
@cmsbuild please test

**Comment by jpata** [comment]
is this intended for 12_4 as well? 😨

**Comment by mbluj** [comment]
> is this intended for 12_4 as well? fearful

It will be great, but it is less critical than new deepTau. One the other hand it is (hopefully) also simpler.

**Comment by jpata** [comment]
Did you check the AOD and MINIAOD size increase on a large enough sample (e.g. 1k)? Runtime CPU/mem cost?
LogMessage is preferred to cout. FIXMEs should be issues.

**Comment by mbluj** [comment]
> Did you check the AOD and MINIAOD size increase on a large enough sample (e.g. 1k)? Runtime CPU/mem cost? LogMessage is preferred to cout. FIXMEs should be issues.

The size increase is small, I will run standard tool and then attach a report;
I will change FIXME to something "softer";
Verbosity in the new module was mainly for initial tests, but I agree LogMessage is more appriopriate I will correct it quickly.

**Comment by jpata** [comment]
assign xpog

(for changes to MINI)

**Comment by jpata** [comment]
type tau

**Comment by jpata** [comment]
@cmsbuild please test

**Comment by mbluj** [comment]
For completeness timing report (400 ttbar events generated as for size checks, but with `--customize Validation/Performance/TimeMemoryInfo.py` added, and RECO and PAT steps separated):
* REDO/AOD
```
bash ~/tools/timeDiffFromReport.sh ~/step4_RECO_TimeMemoryInfo.log ~/step4_RECO_TimeMemoryInfo_mod.log
Couldn't parse time report using CPU and wall-clock format: trying Wall-clock only
  delta/mean delta/orJob     original                   new       module name
  ---------- ------------     --------                  ----       ------------
   +1.000000      +0.00%         0.00 ms/ev ->         0.01 ms/ev muidTMLastStationLoose
  ---------- ------------     --------                  ----       ------------
Job total:  6.38503 s/ev ==> 6.42496 s/ev

The same excluding the first 1 events
  ---------- ------------     --------                  ----       ------------
Job total:  6.36961 s/ev ==> 6.40943 s/ev
```
=> negligible (biggest difference in module not related to this PR which is below threshold when 1st event excluded)

* MiniAOD
```
bash ~/tools/timeDiffFromReport.sh ~/step5_PAT_TimeMemoryInfo.log  ~/step5_PAT_TimeMemoryInfo_mod.log
Couldn't parse time report using CPU and wall-clock format: trying Wall-clock only
  delta/mean delta/orJob     original                   new       module name
  ---------- ------------     --------                  ----       ------------
   +1.162162      +0.03%         0.03 ms/ev ->         0.12 ms/ev slimmedTausNoDeepIDs
   +1.000000      +0.00%         0.00 ms/ev ->         0.01 ms/ev manystripclus53X
   +0.072176      +0.28%        12.75 ms/ev ->        13.71 ms/ev prunedGenParticlesWithStatusOne
   -0.050856      -0.15%        10.48 ms/ev ->         9.96 ms/ev chargedHadronTrackResolutionFilter
  ---------- ------------     --------                  ----       ------------
Job total:  0.340307 s/ev ==> 0.346888 s/ev

The same excluding the first 1 events
  delta/mean delta/orJob     original                   new       module name
  ---------- ------------     --------                  ----       ------------
   +1.174350      +0.03%         0.03 ms/ev ->         0.12 ms/ev slimmedTausNoDeepIDs
   +0.072410      +0.28%        12.69 ms/ev ->        13.64 ms/ev prunedGenParticlesWithStatusOne
  ---------- ------------     --------                  ----       ------------
Job total:  0.335329 s/ev ==> 0.342436 s/ev
```
=> very small: visible increase of timing of modified module `slimmedTausNoDeepIDs` (0.03 -> 0.12 ms/ev), but total impact on miniAOD job is +0.03%

**Comment by mbluj** [comment]
I do not quote explicitly memory use with `~/tools/getTimeMemSummary.sh`, but noting unexpected noticed.

**Comment by jpata** [comment]
Thanks for these. I agree it looks reasonable. Let's wait for the test results.

**Comment by jpata** [comment]
@smuzaffar is everything OK with the tests here? they look stuck / not started.

**Comment by perrotta** [comment]
please test
(let retry)

**Comment by smuzaffar** [comment]
looks like issue with baseline generation ( job was hanging for for last 8 hours : https://cmssdt.cern.ch/jenkins/job/ib-run-baseline/773/ ). I have killed and restarted the IB baseline  job again

**Comment by mariadalfonso** [comment]
+xpog

(extra information added in the tau, for negligible size increase)

**Comment by mariadalfonso** [comment]
out of curiosity: 
are those extra informations helping to reproduce the TAU-ID from the mini to match those at RECO ?
If yes would be helpful to add some link to previous presentation in the PR's description

**Comment by jpata** [comment]
+reconstruction
- additional links between taus and lost tracks at RECO and MINI level
- no appreciable change to runtime or event size

**Comment by mbluj** [comment]
> out of curiosity: are those extra informations helping to reproduce the TAU-ID from the mini to match those at RECO ? If yes would be helpful to add some link to previous presentation in the PR's description

No, it is link to store information about charged prongs of taus which are built from tracks not used by pflow. Those tracks, stored in miniAOD as collection of Packedcandidates called lostTracks, are already used by tau-reco for miniAOD as replacement of standard tracks used at RECO/AOD.

**Comment by mbluj** [comment]
This PR was recently fully signed, but in meantime milestone was changed from 12_4_X to 12_5_X. Should we backport this development to have it in 12_4_X? (Technically it will not be a real backport, but a new PR from same dev branch to different target branch.)

**Comment by jpata** [comment]
Backports for physics-changing quantities are not allowed, unless only enabled with a modifier.

**Comment by perrotta** [comment]
@mbluj CMSSW_12_4_0_pre4 was built a few hours ago, and it being the last open pre-release in the 12_4_X cycle the development has now moved to 12_5_X.

Technically, if you want to backport anything to 12_4_X from the development cycle a backport PR is needed.
Please keep in mind the rules for the backports after the last open pre-release (and before the final release), which for reco PRs basically reads "only bug fixes, or changes that do not modify the reco output".

This PR was declared to be "nice to have, but not critical for 12_4" at the last ORP meeting, and as such we did not waited for it before closing the pre-release. Being it not critical for 12_4 I understand that you can aim it for 12_5. If you want to backport in 12_4 to have it available for some test in that release, the modifier switch suggested by @jpata is also an option.

**Comment by mbluj** [comment]
OK, it is indeed not critical and I know rules. I just hoped that it can be still enter to 124X as was accepted ~15' after switch of target release.

File: RecoTauTag/RecoTau/plugins/PFRecoTauLostTrackPlugin.cc [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+#include "DataFormats/TauReco/interface/PFRecoTauChargedHadron.h"
+#include "DataFormats/TauReco/interface/PFRecoTauChargedHadronFwd.h"
+
+namespace reco {
+  namespace tau {
+
+    class PFRecoTauLostTrackPlugin : public RecoTauModifierPlugin {
+    public:
+      explicit PFRecoTauLostTrackPlugin(const edm::ParameterSet&, edm::ConsumesCollector&& iC);
+      ~PFRecoTauLostTrackPlugin() override;
```
Comment (perrotta) [inline RecoTauTag/RecoTau/plugins/PFRecoTauLostTrackPlugin.cc:28]:
```suggestion
      ~PFRecoTauLostTrackPlugin() override = default;
```

File: RecoTauTag/RecoTau/plugins/PFRecoTauLostTrackPlugin.cc [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+      void operator()(PFTau&) const override;
+      void beginEvent() override;
+      void endEvent() override;
+
+    private:
+      edm::Handle<reco::TrackCollection> tracks_;
+      edm::EDGetTokenT<reco::TrackCollection> track_token_;
+      int verbosity_;
```
Comment (perrotta) [inline RecoTauTag/RecoTau/plugins/PFRecoTauLostTrackPlugin.cc:36]:
```suggestion
      const edm::EDGetTokenT<reco::TrackCollection> track_token_;
      const int verbosity_;
```
and initialize them in the class initialization list

File: RecoTauTag/RecoTau/plugins/PFRecoTauLostTrackPlugin.cc [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    };
+
+    PFRecoTauLostTrackPlugin::PFRecoTauLostTrackPlugin(const edm::ParameterSet& cfg, edm::ConsumesCollector&& iC)
+        : RecoTauModifierPlugin(cfg, std::move(iC)) {
+      track_token_ = iC.consumes(cfg.getParameter<edm::InputTag>("trackSrc"));
+      verbosity_ = cfg.getParameter<int>("verbosity");
+    }
+
+    PFRecoTauLostTrackPlugin::~PFRecoTauLostTrackPlugin() {}
```
Comment (perrotta) [inline RecoTauTag/RecoTau/plugins/PFRecoTauLostTrackPlugin.cc:45]:
```suggestion
```
(remove if you define it as `default` in L28)

File: RecoTauTag/RecoTau/plugins/PFRecoTauLostTrackPlugin.cc [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    private:
+      edm::Handle<reco::TrackCollection> tracks_;
+      edm::EDGetTokenT<reco::TrackCollection> track_token_;
+      int verbosity_;
+    };
+
+    PFRecoTauLostTrackPlugin::PFRecoTauLostTrackPlugin(const edm::ParameterSet& cfg, edm::ConsumesCollector&& iC)
+        : RecoTauModifierPlugin(cfg, std::move(iC)) {
+      track_token_ = iC.consumes(cfg.getParameter<edm::InputTag>("trackSrc"));
+      verbosity_ = cfg.getParameter<int>("verbosity");
```
Comment (perrotta) [inline RecoTauTag/RecoTau/plugins/PFRecoTauLostTrackPlugin.cc:42]: This should be "untracked"

File: RecoTauTag/RecoTau/plugins/PFRecoTauLostTrackPlugin.cc [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+ *
+ * =============================================================================
+ */
+
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+
+#include "RecoTauTag/RecoTau/interface/RecoTauBuilderPlugins.h"
+#include "DataFormats/TauReco/interface/PFTau.h"
+#include "DataFormats/TauReco/interface/PFRecoTauChargedHadron.h"
+#include "DataFormats/TauReco/interface/PFRecoTauChargedHadronFwd.h"
```
Comment (perrotta) [inline RecoTauTag/RecoTau/plugins/PFRecoTauLostTrackPlugin.cc:20]: This does not seem to be used and can be removed, unless you want to use the definition of `PFRecoTauChargedHadronCollection` at [L59](https://github.com/cms-sw/cmssw/pull/37918/files#diff-51256b8c93e7a78492ef9842cc03a30a51d35de4ad31c02f0ff1e9749cff2ccdR59)

**Review by perrotta — COMMENTED** [review]
Some possible code cleanup...

**Comment by mbluj** [comment]
@perrotta thank you. I'll implement proposed improvements.

File: RecoTauTag/RecoTau/plugins/PFRecoTauLostTrackPlugin.cc (hunk 4 above)
Comment (mbluj) [reply RecoTauTag/RecoTau/plugins/PFRecoTauLostTrackPlugin.cc:42]: In principle yes, but in plugin calling this and similar ones it is the "verbose" is defined as tracked. In a reault this small change in this plugin will propagate to other plugins which I think is not worth an effort. Therefore, I will keep it as it is.

**Comment by jpata** [comment]
@cmsbuild please test

**Comment by jpata** [comment]
+reconstruction
- since the last reco sig in https://github.com/cms-sw/cmssw/pull/37918#issuecomment-1127394289, small technical cleanups

**Comment by mbluj** [comment]
@mariadalfonso, I think it should be signed again by xpog after small technical cleanups. Thanks!

**Comment by mariadalfonso** [comment]
+xpog

(extra information added in the tau, for negligible size increase)

**Comment by perrotta** [comment]
+1

**Pull request merged**